fix: accept 0 as a valid value for select fields in perform_add()

jdpaniagua opened the pull request:
Some forms like Package Offer use indexed select lists for fields such as `packagetype` and `billingtype`, where 0 is a valid option.

Currently, these values are skipped during insert because `empty(0)` and `empty("0")` return true in PHP, causing db errors like if the field is defined as NOT NULL.

This change replaces `!empty()` with a more precise check using `isset()` and comparison against an empty string. This ensures that legitimate values such as 0 are preserved and inserted.

- Fixes insertion errors for valid forms using 0-indexed selects
- Maintains compatibility with current behavior for truly empty fields
- Improves overall robustness for form submissions with numeric values

Fixes #19

jdpaniagua committed:
fix: accept 0 as a valid value for select fields in perform_add()

Some forms like Package Offer use indexed select lists for fields such as `packagetype` and `billingtype`, where 0 is a valid option.

Currently, these values are skipped during insert because `empty(0)` and `empty("0")` return true in PHP, causing db errors like if the field is defined as NOT NULL.

This change replaces `!empty()` with a more precise check using `isset()` and comparison against an empty string. This ensures that legitimate values such as 0 are preserved and inserted.

- Fixes insertion errors for valid forms using 0-indexed selects
- Maintains compatibility with current behavior for truly empty fields
- Improves overall robustness for form submissions with numeric values

miken32 commented:
Everywhere else in that loop assumes that `$processed[$field]` is set. It would make more sense to assume that here as well, and do the `isset()` check at the start of the loop. It looks like I just did that for `perform_edit()` a couple of weeks ago in bfc187d7

miken32 commented:
Also it's probably worth asking if there is any problem with adding an empty string to a field? If a field is left blank by the user and it doesn't have a default value, they will get the same error you reported. Could we just skip that check altogether?

jdpaniagua commented:
Yes, I agree — moving the `isset()` check to the top of the loop makes more sense, and I’ll align the logic with `perform_edit()` as in commit `bfc187d`.

Regarding the idea of removing the empty string check entirely: I’m not fully sure yet if that could lead to unexpected results. For instance, in some cases, inserting `NULL` into the database might be intentional. Although I suppose that's what the `check_empty = NO-NULL` flag is meant to enforce.

Also, while reviewing this, I noticed that `perform_add()` only evaluates `check_empty` if a `validator` is defined, whereas `perform_edit()` checks `check_empty` regardless. I wonder if that difference might be unintentional — do you think that’s worth aligning?

miken32 commented:
>in some cases, inserting NULL into the database might be intentional

Possible, but my assumption with this code is that very little is intentional.

As an example: this `check_empty` stuff. The original code did not look for `NO-NULL` in add, only in edit: https://github.com/Star2Billing/a2billing/blob/develop/common/lib/Form/Class.FormHandler.inc.php#L1654. Was that intentional? Who knows. 

Really it should be two boolean parameters `$validate_empty` and `$insert_nulls` or something, rather than a string.

jdpaniagua commented:
I had a look at the original `Class.FormHandler.inc.php`, and I think there might have been a rationale behind the different handling in `perform_edit()`.

My impression is that `perform_edit()` was written to avoid overwriting `NULL` values from the database with empty strings — which would happen if a user left a field untouched in the form. So that extra check made sense there.

In contrast, `perform_add()` didn't need to handle that case explicitly. If a field was left blank in the form, it simply wasn’t included in the insert statement, and the database would apply the default — which was often `NULL`.

So while the code wasn't always clean or abstracted, I think it did reflect the intent to let MySQL handle defaults appropriately during inserts.

miken32 commented:
That's entirely possible. Relying on NULL values in a database is a bad idea regardless, especially when NULL means something different than an empty string. IMO null values should really only be used in cases where a foreign key association is missing. Anyway, I'll merge this once the code is updated per your comment.

Going to look next at updating the agent and customer UI to bootstrap, and getting rid of the epayment stuff which I assume is hopelessly broken.

jdpaniagua committed:
refactor: refine handling of empty values in perform_add()

Builds on a9e5deb to improve validation flow:
- Aligns empty value logic more closely with perform_edit()
- Applies check_empty only when no validator is defined
- Converts "" to null for NO-NULL fields

jdpaniagua commented:
Thanks — I share your view on the use of `NULL`; it's often problematic when it carries semantic meaning. I’ve tried to keep the changes minimal to avoid breaking anything unintentionally.

I've just pushed the updates. Feel free to make any further changes to the PR if you think there's a cleaner or better way to do something — I'd appreciate it, honestly.

miken32 commented on src/Forms/FormHandler.php:
```diff
@@ -1495,37 +1495,46 @@ public function perform_add(string &$form_action): void
+            } elseif ($value === "") {
+                if ($check_empty === "NO-NULL") {
+                    $value = null;
+                } elseif ($check_empty === "NO") {
```
This check for $check_empty === "NO" has to happen *before* the check for the validator. If that element starts with "NO" *and* the field is empty, then we shouldn't run the validator check at all.

jdpaniagua commented:
That makes sense — I initially approached it the other way around, thinking that if a `validator` is present, the field is implicitly required, since an empty value would cause the validation to fail anyway.

But I see your point: it’s more predictable to short-circuit validation when the field is already invalid due to `check_empty`. That said, I wonder if the real issue is that `check_empty` is acting like a validator, but outside the validation system.

Maybe it would be cleaner long-term to replace it with something like a built-in validator — for example, `no-empty` or `empty-is-null` — so that all validation logic, including requiredness and conversion, is handled uniformly through validators.

This would reduce special-case logic in `perform_add()` and `perform_edit()`, and make the validation flow easier to reason about and extend.

In the meantime, happy to adjust the logic per your direction.

miken32 commented:
> if a validator is present, the field is implicitly required, since an empty value would cause the validation to fail anyway

That's the reason for "NO" and "NO-NULL": to indicate that the field *isn't* implicitly required.

> it’s more predictable to short-circuit validation when the field is already invalid due to `check_empty`

The `check_empty` element can't make a field invalid, it makes the field valid – if it's empty.

>Maybe it would be cleaner long-term to replace it with something like a built-in validator

Agreed; in terms of working within the framework we have now, I think the cleanest approach would be:
* find the input elements that use "NO" (about 80 of them) and "NO-NULL" (about 15)
* see which validator callbacks they use
* update those validators to allow for empty strings as well
* * this could be an addition to the existing class, e.g. `Validator::min3CharsOrEmpty()`
* * or creation of a new class `OptionalValidator::min3Chars()`
* remove the check_empty parameter, and add a boolean one to catch the elements that are using "NO-NULL"

For the purposes of this PR though, I think just doing the `array_key_exists()` check for `$processed[$field]` at the top of the loop, and then replacing the `empty()` check with an empty string check should suffice.

jdpaniagua committed:
fix: handle 0 and empty string values correctly in perform_add()

Builds on previous changes to fully align with perform_edit():
- Uses array_key_exists() to ensure $processed[$field] is present
- Replaces empty() with explicit comparison to ""

jdpaniagua commented:
That's a pretty embarrassing oversight — I had understood the logic of `check_empty` almost exactly backwards. Thanks for the clarification and your patience.

I'm updating the PR now to follow the validation flow in `perform_edit()`, including the ordering of `check_empty` and validator calls.

miken32 commented on src/Forms/FormHandler.php:
```diff
@@ -1510,22 +1511,26 @@ public function perform_add(string &$form_action): void
                     if ($result !== true) {
                         $this->all_fields_valid = false;
-                        $form_action = "ask-add";
+                        $form_action = "ask-edit";
```
Guessing this was a copy/paste error. Tried to update but didn't have permissions to edit the PR. I'll fix after merge.

miken32 commented on src/Forms/FormHandler.php:
```diff
@@ -1510,22 +1511,26 @@ public function perform_add(string &$form_action): void
             // CHECK IF THIS IS A SPLITABLE FIELD LIKE 012-014 OR 15,16,17
             if (in_array($field, $this->FG_SPLITABLE_FIELDS)) {
-                $value = $processed[$field];
                 if (empty($value) || str_starts_with($value, "_")) {
```
Same with this one

jdpaniagua replied:
You're absolutely right — that was an oversight on my part, and I really appreciate you catching and fixing it after the merge.

Sorry for taking up so much of your time over something that should’ve been trivial to get right. I’ll be more careful on future contributions.

miken32 replied:
No worries, I guarantee you can find 2 dozen cases of me breaking something terribly in this codebase and then fixing it days or months later, especially in the web UI which we don't use that much here. (I am being a little more careful now, knowing I'm not the only one using it lol)
